From cea50042f6d697a668ed8c00934dd597905a05c9 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <michael.i.doherty@intel.com>
Date: Thu, 24 Sep 2015 19:15:55 +0100
Subject: [PATCH 14/42] kernel-install: Support alternate root usage via
 SUBDIR/PLUGIN_SUBDIR

This enables kernel-install to be used with an alternative root, by accepting a new
--root argument. Supporting plugins can make use of the PLUGIN_SUBDIR variable
in the environment to prefix all of their operations with the appropriate root.

This ensures that only the files from the given root's filesystem tree are
used, reducing any host contamination risks or even relying on software being
present on the host system, during image/filesystem generations.

Signed-off-by: Ikey Doherty <michael.i.doherty@intel.com>
---
 src/kernel-install/kernel-install | 61 +++++++++++++++++++++++--------
 1 file changed, 46 insertions(+), 15 deletions(-)

diff --git a/src/kernel-install/kernel-install b/src/kernel-install/kernel-install
index c9a80b2e3d..06b1f3a860 100755
--- a/src/kernel-install/kernel-install
+++ b/src/kernel-install/kernel-install
@@ -28,8 +28,11 @@ usage()
     echo "Options:"
     echo "  -h,--help     Print this help"
     echo "  -v,--verbose  Increase verbosity"
+    echo "  -p,--root     Optional. Prefix kernel-install operations with different root"
 }
 
+SUBDIR=""
+
 dropindirs_sort()
 {
     local suffix=$1; shift
@@ -58,12 +61,26 @@ dropindirs_sort()
 
 export LC_COLLATE=C
 
-for i in "$@"; do
-    if [ "$i" == "--help" -o "$i" == "-h" ]; then
+args=("$@")
+for ((i=0; i < $#; i++)) {
+    arg="${args[$i]}"
+    if [ "$arg" == "--help" -o "$arg" == "-h" ]; then
         usage
         exit 0
+    elif [ "$arg" == "--root" ]; then
+        if [[ "$((i+1))" -lt $# ]]; then
+            SUBDIR="${args[$((i+1))]}"
+            export SUBDIR
+        else
+            usage
+            exit 1
+        fi
     fi
-done
+}
+
+if [[ ! -z "${SUBDIR}" ]]; then
+    shift; shift;
+fi
 
 KERNEL_INSTALL_VERBOSE=0
 if [ "$1" == "--verbose" -o "$1" == "-v" ]; then
@@ -86,12 +103,24 @@ KERNEL_VERSION="$1"
 KERNEL_IMAGE="$2"
 
 # Reuse directory created without a machine ID present if it exists.
-if [[ -d /efi/Default ]] || [[ -d /boot/Default ]] || [[ -d /boot/efi/Default ]]; then
+if [[ -d ${SUBDIR}/efi/Default ]] || [[ -d ${SUBDIR}/boot/Default ]] || [[ -d ${SUBDIR}/boot/efi/Default ]]; then
     MACHINE_ID="Default"
-elif [[ -f /etc/machine-id ]]; then
-    read MACHINE_ID < /etc/machine-id
+elif [[ -f ${SUBDIR}/etc/machine-id ]]; then
+    read MACHINE_ID < ${SUBDIR}/etc/machine-id
+fi
+
+if ! [[ $MACHINE_ID ]]; then
+    echo "Could not determine your machine ID from ${SUBDIR}/etc/machine-id." >&2
+    echo "Please run 'systemd-machine-id-setup' as root. See man:machine-id(5)" >&2
+    if [[ -z "${SUBDIR}" ]]; then
+        exit 1
+    else
+        # Enable bare-chroot usage.
+        BOOT_DIR_ABS="${SUBDIR}/boot"
+    fi
 else
     MACHINE_ID="Default"
+    BOOT_DIR_ABS="${SUBDIR}/boot/$MACHINE_ID/$KERNEL_VERSION"
 fi
 
 if [[ ! $COMMAND ]] || [[ ! $KERNEL_VERSION ]]; then
@@ -119,8 +148,8 @@ ret=0
 
 readarray -t PLUGINS <<<"$(
     dropindirs_sort ".install" \
-        "/etc/kernel/install.d" \
-        "/usr/lib/kernel/install.d"
+        "${SUBDIR}/etc/kernel/install.d" \
+        "${SUBDIR}/usr/lib/kernel/install.d"
 )"
 
 case $COMMAND in
@@ -138,8 +167,8 @@ case $COMMAND in
         for f in "${PLUGINS[@]}"; do
             if [[ -x $f ]]; then
                 [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && \
-                    echo "+$f add $KERNEL_VERSION $ENTRY_DIR_ABS $KERNEL_IMAGE ${INITRD_OPTIONS[@]}"
-                "$f" add "$KERNEL_VERSION" "$ENTRY_DIR_ABS" "$KERNEL_IMAGE" "${INITRD_OPTIONS[@]}"
+                    echo "+PLUGIN_SUBDIR=${SUBDIR} $f add $KERNEL_VERSION $BOOT_DIR_ABS $KERNEL_IMAGE ${INITRD_OPTIONS[@]}"
+                PLUGIN_SUBDIR="${SUBDIR}" "$f" add "$KERNEL_VERSION" "$BOOT_DIR_ABS" "$KERNEL_IMAGE" "${INITRD_OPTIONS[@]}"
                 x=$?
                 if [[ $x == $SKIP_REMAINING ]]; then
                     break
@@ -153,8 +182,8 @@ case $COMMAND in
         for f in "${PLUGINS[@]}"; do
             if [[ -x $f ]]; then
                 [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && \
-                    echo "+$f remove $KERNEL_VERSION $ENTRY_DIR_ABS"
-                "$f" remove "$KERNEL_VERSION" "$ENTRY_DIR_ABS"
+                    echo "+PLUGIN_SUBDIR={SUBDIR} $f remove $KERNEL_VERSION $ENTRY_DIR_ABS"
+                PLUGIN_SUBDIR="${SUBDIR}" "$f" remove "$KERNEL_VERSION" "$BOOT_DIR_ABS"
                 x=$?
                 if [[ $x == $SKIP_REMAINING ]]; then
                     break
@@ -163,10 +192,12 @@ case $COMMAND in
             fi
         done
 
-        [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && \
-            echo "Removing $ENTRY_DIR_ABS"
 
-        rm -rf "$ENTRY_DIR_ABS"
+        if [[ -z "${SUBDIR}" ]]; then
+            [ "$KERNEL_INSTALL_VERBOSE" -gt 0 ] && \
+                echo "Removing $ENTRY_DIR_ABS"
+            rm -rf "$BOOT_DIR_ABS"
+        fi
         ((ret+=$?))
         ;;
 
-- 
2.32.0

